lustre/ws.lua

```diff
@@ -0,0 +1,351 @@
+    req:add_header('Sec-Websocket-Version', 13)
+    req:add_header('Sec-Websocket-Key', self.handshake_key)
+    req:add_header("Host", string.format("%s:%d", host, port))
+    for _, prot in ipairs(self.config.protocols) do
+        --TODO I think luncheon should be able to handle multiple values, but
+        -- it currently only sends the last value added
+        req:add_header("Sec-Websocket-Protocol", prot)
+    end
+    local s, err = req:send()
+    if not s then
+        return "handshake request failure: "..err
+    end
+    local res, err = Response.tcp_source(self.socket)
+    if not res then
+        return nil, "Upgrade response failure: "..err
+    end
+    local handshake = Handshake.client(self.handshake_key, {}, {})
+    local success, err = handshake:validate_accept(res)
+    if not success then
+        return nil, 'invalid handshake: '..err
+    end
+    cosock.spawn(function() self:receive_loop() end, 'Client receive loop')
+    return 1
+end
+
+function WebSocket:accept()
+end
+
+---@param close_code CloseCode
+---@param reason string
+---@return err string|nil
+function WebSocket:close(close_code, reason)
+    local close_frame = Frame.close(close_code, reason):set_mask()
+    local sent, err = self.socket:send(close_frame:encode())
+    if not sent then return nil, err end
+    print("SENT CLOSE FRAME: \n", utils.table_string(close_frame), "\n\n")
+    self._close_frame_sent = true
+    if self.close_cb then self.close_cb(reason) end
+    return 1
+end
+
+---@return message Message
+---@return err string|nil
+--TODO break out some helper functions for this...
+--TODO confirm the socket lifecycle throughout the receive loop: the receive loop ends when the socket is closed
+function WebSocket:receive_loop()
+    local partial_frames = {}
+    local received_bytes = 0
+    local frames_since_last_ping = 0
+    local pending_pong = false
+    local multiframe_message = false 
+    local msg_type
+    while true do
+        local recv, _, err = socket.select({self.socket, self._rx}, nil, self.config._keep_alive)
+        if not recv then
+            if err == "timeout" then
+                self.socket:send(Frame.ping():set_mask():encode())
+                pending_pong = true
+            end
+            goto continue
+        end
+        if recv[1] == self.socket then
+            local frame, err = Frame.from_stream(self.socket)
+            if not frame then
+                if self._close_frame_sent then
+                    --TODO this error case is a little weird, but it seems
+                    -- needed atm to ensure the loop ends when the server initates a close
+                    print("weird error case exit loop: \n", self.socket:getpeername())
+                    return
+                elseif err == "invalid opcode" or err == "invalid rsv bit" then
+                    self:close(CloseCode.protocol(), err)
+                elseif err == "timeout" and self.error_cb then
```

Comment:
We may want to add a todo here since timeout means we should be re-trying to complete the pending frame. I think treating it like an error for now is ok. We should probably include the partial frame in the callback arguments